sql: add assignment cast for UPDATEs

mgartner opened the pull request:
#### opt: fix FK WithScan column type

Previously, the WithScan columns of cascade FK updates were given the
type of the child table's target column. This was incorrect when the
parent column type did not match. This commit fixes the issue by given
the WithScan columns the type of the buffered columns used to update the
parent relation.

This did not cause any known bugs, but it is required to implement
assignment casts with FK cascades.

Release note: None

#### sql: add assignment cast for UPDATEs

Fixes #70628

Release note: None

#### opt: give synthesized assignment cast columns descriptive names

Columns synthesized for assignment casts now have more descriptive
metadata names in the form `<target_column_name>_cast`. This is purely
an aesthetic change and has no effect on semantics.

Release note: None

mgartner committed:
opt: fix FK WithScan column type

Previously, the WithScan columns of cascade FK updates were given the
type of the child table's target column. This was incorrect when the
parent column type did not match. This commit fixes the issue by given
the WithScan columns the type of the buffered columns used to update the
parent relation.

This did not cause any known bugs, but it is required to implement
assignment casts with FK cascades.

Release note: None

mgartner committed:
sql: add assignment cast for UPDATEs

Fixes #70628

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74180)
<!-- Reviewable:end -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 22 of 22 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-:-Ms5pq4ZCIp0MkQIRknt:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach, @mgartner, and @RaduBerinde)

___
*[pkg/sql/opt/optbuilder/scope.go, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-Ms5knX5CuuJIr-NOSXn:-Ms5knX6EDY7h55N6jCv:b-jzxdpm) ([raw file](https://github.com/cockroachdb/cockroach/blob/af62616ca2d76fc1a390fcf88534288f72e3d321/pkg/sql/opt/optbuilder/scope.go#L320)):*
> ```Go
> }
> 
> // getColumn returns the scopeColumn with the given id and reference name
> ```

getColumn -> getColumnWithIDAndreferenceName
___
*[pkg/sql/logictest/testdata/logic_test/cast, line 384 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-Ms5l_5a1NGMZ-H07X5-:-Ms5l_5a1NGMZ-H07X50:beq4nah) ([raw file](https://github.com/cockroachdb/cockroach/blob/af62616ca2d76fc1a390fcf88534288f72e3d321/pkg/sql/logictest/testdata/logic_test/cast#L384)):*
> ```
> EXECUTE update_c('foo'::STRING)
> 
> statement
> ```

should this be `statement ok`?
___
*[pkg/sql/logictest/testdata/logic_test/cast, line 392 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-Ms5lc3g4FWYWc754gaM:-Ms5lc3g4FWYWc754gaN:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/af62616ca2d76fc1a390fcf88534288f72e3d321/pkg/sql/logictest/testdata/logic_test/cast#L392)):*
> ```
> ""
> 
> statement
> ```

ditto
___
*[pkg/sql/logictest/testdata/logic_test/cast, line 409 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-Ms5mRVS-5Fsz1drsaix:-Ms5mRVS-5Fsz1drsaiy:b9l9k0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/af62616ca2d76fc1a390fcf88534288f72e3d321/pkg/sql/logictest/testdata/logic_test/cast#L409)):*
> ```
> 
> statement error \"char\" out of range
> UPDATE assn_cast SET qc = 1234
> ```

why does this one fail while the previous one succeeds? (maybe add a comment)
___
*[pkg/sql/opt/optbuilder/testdata/partial-indexes, line 542 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-Ms5obrG1jzW8V4XrROv:-Ms5obrG1jzW8V4XrROw:bw9yi4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/af62616ca2d76fc1a390fcf88534288f72e3d321/pkg/sql/opt/optbuilder/testdata/partial-indexes#L542)):*
> ```
> # Regression test for #61520. The new value for column a should be in-scope when
> # building the partial index PUT expression when the value in the FROM clause
> # must assignment cast.
> ```

nit: must assignment cast -> must be assignment cast ?
___
*[pkg/sql/opt/xform/testdata/external/trading-mutation, line 1595 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-Ms5pc6JEAlL-A2KZe2p:-Ms5pc6JEAlL-A2KZe2q:bp7bhfm) ([raw file](https://github.com/cockroachdb/cockroach/blob/af62616ca2d76fc1a390fcf88534288f72e3d321/pkg/sql/opt/xform/testdata/external/trading-mutation#L1595)):*
> ```
>       │              └── q:37
>       └── projections
>            ├── assignment-cast: DECIMAL(10,4) [as=column53:53, outer=(23,25), immutable]
> ```

can we give the new column a friendly name? (e.g., just keep the name `buyprice` here)


<!-- Sent from Reviewable.io -->

mgartner committed:
opt: give synthesized assignment cast columns descriptive names

Columns synthesized for assignment casts now have more descriptive
metadata names in the form `<target_column_name>_cast`. This is purely
an aesthetic change and has no effect on semantics.

Release note: None

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74180)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/optbuilder/scope.go, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-Ms5knX5CuuJIr-NOSXn:-MsAzY-G-qpGo5lJ6Y5_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af62616ca2d76fc1a390fcf88534288f72e3d321/pkg/sql/opt/optbuilder/scope.go#L320)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

getColumn -> getColumnWithIDAndreferenceName
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/cast, line 384 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-Ms5l_5a1NGMZ-H07X5-:-MsAznCZB1NCQYhXSOUJ:b-o7vuhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/af62616ca2d76fc1a390fcf88534288f72e3d321/pkg/sql/logictest/testdata/logic_test/cast#L384)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should this be `statement ok`?
</blockquote></details>

Good catch! Done.
___
*[pkg/sql/logictest/testdata/logic_test/cast, line 392 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-Ms5lc3g4FWYWc754gaM:-MsAzpZN6mteo34t2zDP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af62616ca2d76fc1a390fcf88534288f72e3d321/pkg/sql/logictest/testdata/logic_test/cast#L392)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/cast, line 409 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-Ms5mRVS-5Fsz1drsaix:-MsB5GWk0VRgN6-Q4c3G:b-y046sf) ([raw file](https://github.com/cockroachdb/cockroach/blob/af62616ca2d76fc1a390fcf88534288f72e3d321/pkg/sql/logictest/testdata/logic_test/cast#L409)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

why does this one fail while the previous one succeeds? (maybe add a comment)
</blockquote></details>

Done. integer to "char" casts convert the integer to the corresponding 7-bit ASCII character, so anything greater than 127 is out of range.
___
*[pkg/sql/opt/optbuilder/testdata/partial-indexes, line 542 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-Ms5obrG1jzW8V4XrROv:-MsB5ZxoET8nYpXWSlyC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af62616ca2d76fc1a390fcf88534288f72e3d321/pkg/sql/opt/optbuilder/testdata/partial-indexes#L542)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: must assignment cast -> must be assignment cast ?
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/testdata/external/trading-mutation, line 1595 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-Ms5pc6JEAlL-A2KZe2p:-MsBFFCn6fJ0O_XEfoWD:b-jqt369) ([raw file](https://github.com/cockroachdb/cockroach/blob/af62616ca2d76fc1a390fcf88534288f72e3d321/pkg/sql/opt/xform/testdata/external/trading-mutation#L1595)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

can we give the new column a friendly name? (e.g., just keep the name `buyprice` here)
</blockquote></details>

I added a new commit which gives them names in the form `<target_column_name>_cast`.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 22 of 22 files at r3, 22 of 22 files at r4, 20 of 20 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74180#-:-MsBLnLaBKXDQjGo5Lz8:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach and @RaduBerinde)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTR!

bors r+

mgartner commented:
bors r+
